[BEAM-1418] MapElements and FlatMapElements should comply with PTransform style guide

jkff opened the pull request:
For both of these classes, we change (Flat)MapElements.via(fn).withOutputType(td) to (Flat)MapElements.into(td).via(fn), which is both shorter and allows getting rid of the ugly intermediate and publicly-visible class MissingOutputTypeDescriptor.

R: @bjchambers

jkff commented:
Ping.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/MapElements.java:
```diff
@@ -26,14 +26,34 @@
+  private final DisplayData.ItemSpec<?> fnClassDisplayData;
+
+  private MapElements(
+      SimpleFunction<InputT, OutputT> fn, TypeDescriptor<OutputT> outputType, Class<?> fnClass) {
```
`@Nullable` where appropriate

jkff replied:
Done, and explained the nullness nature of all fields.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/FlatMapElements.java:
```diff
@@ -29,6 +29,29 @@
+  private final transient TypeDescriptor<Iterable<OutputT>> outputType;
+  private final DisplayData.ItemSpec<?> fnClassDisplayData;
+
+  private FlatMapElements(
```
`@Nullable` where appropriate

jkff replied:
Done

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/FlatMapElements.java:
```diff
@@ -82,50 +100,7 @@
+    return new FlatMapElements(fn, null, fn.getClass());
   }
 
   @Override
```
This may need some validation to make sure it is fully specified prior to expand time.

jkff replied:
Done

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/MapElements.java:
```diff
@@ -77,37 +92,7 @@
+    return new MapElements<>(fn, null, fn.getClass());
   }
 
   @Override
```
Ditto

jkff replied:
Done

jkff commented:
Ping.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/FlatMapElements.java:
```diff
@@ -82,54 +82,45 @@
    */
   public static <InputT, OutputT> FlatMapElements<InputT, OutputT>
```
Also, in the interest of standardizing on `into` first, why not make `via` a non static method?

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/FlatMapElements.java:
```diff
@@ -82,54 +82,45 @@
+  via(SerializableFunction<NewInputT, ? extends Iterable<OutputT>> fn) {
+    return new FlatMapElements(
+        SimpleFunction.fromSerializableFunctionWithOutputType(fn, (TypeDescriptor) outputType),
+        null,
```
shouldn't this preserve the existing type descriptor, if any?

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/MapElements.java:
```diff
@@ -77,41 +73,46 @@
    */
   public static <InputT, OutputT> MapElements<InputT, OutputT> via(
```
Same comments re: making this non-static and/or inferring the outputType from the fn.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/MapElements.java:
```diff
@@ -77,41 +73,46 @@
 
   @Override
   public PCollection<OutputT> expand(PCollection<? extends InputT> input) {
+    checkNotNull(fn, "Must specify a function on MapElements using .via()");
```
Is it worth making this require the output type is specified?

jkff commented:
Both before and after this PR, there's two ways to construct a MapElements or FlatMapElements:
1. Construct from SerializableFunction and a TypeDescriptor
2. Construct from a SimpleFunction (which has its own TypeDescriptor)

In both cases, internally the transform stores the SimpleFunction, but in case 2 it just uses what the user gave directly, and in case 1 it instead constructs the SimpleFunction from the SerializableFunction and the TypeDescriptor.

This PR leaves case 2 unchanged - static `via(SimpleFunction)`.
It syntactically changes case 1 from `.via(SerializableFunction).withOutputType(TypeDescriptor)` into `.into(TypeDescriptor).via(SerializableFunction)`, with the sole motivation of getting rid of the ugly and user-visible intermediate class. The TypeDescriptor private field is used only as an intermediate storage to put the TypeDescriptor supplied in .into() in order to combine it into the SimpleFunction when you finish application with .via().

There's no semantic changes, or changes in quality of type inference in this PR, and I would like to keep it this way.

I think this addresses all your comments - let me know if anything remains unclear.

bjchambers commented:
LGTM
